src/valarray.hpp

```diff
@@ -153,4 +191,101 @@ template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto min(q
 	return min_val;
 }
 
+// max(array)
+template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto max(quokka::valarray<T, d> const &v) -> T
+{
+  static_assert(d >= 1);
+  T max_val = v[0]; // v must have at least 1 element
+
+  for (size_t i = 0; i < v.size(); ++i) {
+    max_val = std::max(max_val, v[i]);
+  }
+  return max_val;
+}
+
+// sum(array)
+template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto sum(quokka::valarray<T, d> const &v) -> T
+{
+  T sum_val = 0;
+  for (size_t i = 0; i < v.size(); ++i) {
+    sum_val += v[i];
+  }
+  return sum_val;
+}
+
+// hasnan(array)
+template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto hasnan(quokka::valarray<T, d> const &v) -> bool
+{
+  for (size_t i = 0; i < v.size(); ++i) {
+    if (std::isnan(v[i])) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// array > array
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator>(quokka::valarray<T, d> const &a, quokka::valarray<T, d> const &b) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] > b[i];
+  }
+  return comp;
+}
+
+// array > scalar
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator>(quokka::valarray<T, d> const &a, T const &scalar) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] > scalar;
+  }
+  return comp;
+}
+
+// array < array
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator<(quokka::valarray<T, d> const &a, quokka::valarray<T, d> const &b) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] < b[i];
+  }
+  return comp;
+}
+
+// array < scalar
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator<(quokka::valarray<T, d> const &a, T const &scalar) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] < scalar;
+  }
+  return comp;
+}
+
+// define fillin(valarray) for quokka::valarray
+template <typename T, int d> 
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void fillin(quokka::valarray<T, d> &v, T const &scalar)
```

Comment:
Could you make this a static member function of `quokka::valarray`?

Reply:
You marked it `static`, but it is not a member function of `valarray`. Can you change this to `void valarray::fillin(...)`?

Reply:
Oh, that's what you mean. I was wondering about that since fillin apparently change the content of the array and shouldn't be 'static'.

Reply:
Not quite sure how. I tried `AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void valarray::fillin...` and `AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void quokka::valarray::fillin` but it didnt' work. It complained valarray is not defined, or quokka::valarray used without template arguments. I renamed fillin to valarray_fillin for now to avoid conflicting names. Let me know if you have more suggestions.

Reply:
There are some more examples here: https://learn.microsoft.com/en-us/cpp/cpp/class-templates?view=msvc-170#member-functions-of-class-templates

Reply:
Thanks. I used a different approach. Now fillin is a build in function of valarray. Use it like array.fill(value). Same thing with hasnan. Fully tested and working properly. How do you like it?